### PR TITLE
build-tracker: keep better track of jobs

### PR DESCRIPTION
**Pull request by burmudar**
### The following things have changed and I'd appreciate reviews on:
* Build has moved to its own package, and instead of keeping track individual jobs, the Build now keeps track of Steps.
* Steps accumulate Jobs with the same name. A Step is able to determine its `FinalStatus`, which is done by looking at the last two Jobs to determine whether the step has Passed, Failed or is Fixed.
* All things Notification like has been moved to the notify package. The notifiy client no longer keeps track of teammates, instead, it keeps track of notifications that have been sent for a build. ~If a build wants to send an updated notification, the teammate is reused from the previous build.~ In a follow up PR I will address this more thoroughly.
* The notification client now only works with a `BuildNotification` and has no access to the original build. This makes the boundary between keeping track of a build and sending a notification very clear. The only way to send a notification is to provide all the information in a BuildNotification

## Test plan
Unit Tests and Integrations tests see this channel https://sourcegraph.slack.com/archives/C03MM9C63KQ

### CODE REVIEW

File: dev/build-tracker/build/steps.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+	}
+	if len(s.Jobs) == 1 {
+		return s.LastJob().status()
+	}
+	// we only care about the last two states of because that determines the final state
+	// n - 1  |   n    | Final
+	// Passed | Passed | Passed
+	// Passed | Failed | Failed
+	// Failed | Failed | Failed
+	// Failed | Passed | Fixed
```
Comment (jhchabran) [inline dev/build-tracker/build/steps.go:89]: Thanks for taking the time to write this 🙏

File: go.sum [hunk 2]
```diff
@@ -1364,6 +1364,7 @@ github.com/hashicorp/yamux v0.0.0-20180604194846-3520598351bb/go.mod h1:+NfK9FKe
 github.com/hashicorp/yamux v0.0.0-20181012175058-2f1d1f20f75d/go.mod h1:+NfK9FKeTrX5uv1uIXGdwYDTeHna2qgaIlx54MXqjAM=
 github.com/hexops/autogold v0.8.1/go.mod h1:97HLDXyG23akzAoRYJh/2OBs3kd80eHyKPvZw0S5ZBY=
 github.com/hexops/autogold v1.3.1 h1:YgxF9OHWbEIUjhDbpnLhgVsjUDsiHDTyDfy2lrfdlzo=
+github.com/hexops/autogold v1.3.1/go.mod h1:sQO+mQUCVfxOKPht+ipDSkJ2SCJ7BNJVHZexsXqWMx4=
```
Comment (jhchabran) [inline go.sum:1367]: Autogold v1 is banned now!

**Review by jhchabran — CHANGES_REQUESTED** [review]
Minor nits, looks great. Really appreciate the comments. Requesting changes for autogold, but otherwise, LGTM.

File: dev/build-tracker/build/build.go [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+func (b *Build) IsFinished() bool {
+	switch b.GetState() {
+	case "passed", "failed", "blocked", "canceled":
+		return true
+	default:
+		return false
+	}
+}
+
+func (b *Build) authorName() string {
```
[the remark this answers is not in the document]
Comment (burmudar) [reply dev/build-tracker/build/build.go:93]: Agreed! Will change it

File: go.sum (hunk 2 above)
Comment (burmudar) [reply go.sum]: I did remove the use of it but still see it in the `go.sum` file even after `go mod tidy`

File: dev/build-tracker/build/build.go [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+type Build struct {
+	// Build is the buildkite.Build currently being executed by buildkite on a particular Pipeline
+	buildkite.Build `json:"build"`
+
+	// Pipeline is a wrapped buildkite.Pipeline that is running this build.
+	Pipeline *Pipeline `json:"pipeline"`
+
+	// steps is a map that keeps track of all the buildkite.Jobs associated with this build.
+	// Each step keeps track of jobs associated with that step. Every job is wrapped to allow
+	// for safer access to fields of the buildkite.Jobs. The name of the job is used as the key
```
[the remark this answers is not in the document]
Comment (burmudar) [reply dev/build-tracker/build/build.go:24]: The buildkite job has everything set as a pointer. So the wrapped Job provides "safer" access in it that it won't return null but the default value, like an empty string or 0

File: go.sum (hunk 2 above)
Comment (jhchabran) [reply go.sum]: Strange. Let's try look from where this is coming from.

Comment (burmudar) [reply go.sum]: It comes from our use of `github.com/hexops/valast` in enterprise/cmd/worker/internal/telemetry, so not directly imported by us - see https://github.com/hexops/valast/blob/main/go.mod#L7
```
go mod why "github.com/hexops/autogold"
# github.com/hexops/autogold
github.com/sourcegraph/sourcegraph/enterprise/cmd/worker/internal/telemetry
github.com/sourcegraph/sourcegraph/enterprise/cmd/worker/internal/telemetry.test
github.com/hexops/valast
github.com/hexops/valast.test
github.com/hexops/autogold
```

File: dev/build-tracker/build/build.go [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+	BuildStatusUnknown BuildStatus = ""
+	BuildPassed        BuildStatus = "Passed"
+	BuildFailed        BuildStatus = "Failed"
+	BuildFixed         BuildStatus = "Fixed"
+)
+
+func (b *Build) AddJob(j *Job) error {
+	stepName := j.GetName()
+	if stepName == "" {
+		return fmt.Errorf("job name is empty")
```
[the remark this answers is not in the document]
Comment (burmudar) [reply dev/build-tracker/build/build.go]: Updated

**Review by jhchabran — APPROVED** [review]

**Pull request merged**